src/option/from_stream.rs

```diff
@@ -17,24 +17,26 @@ where
         let stream = stream.into_stream();
 
         Box::pin(async move {
-            // Using `scan` here because it is able to stop the stream early
+            // Using `take_while` here because it is able to stop the stream early
             // if a failure occurs
-            let mut found_error = false;
+            let mut found_none = false;
             let out: V = stream
-                .scan((), |_, elem| {
-                    match elem {
-                        Some(elem) => Some(elem),
-                        None => {
-                            found_error = true;
-                            // Stop processing the stream on error
-                            None
-                        }
+                .take_while(|elem| {
+                    elem.is_some() || {
+                        found_none = true;
+                        // Stop processing the stream on `None`
+                        false
                     }
                 })
+                .map(Option::unwrap)
```

Comment:
This pattern here is quite surprising to me, and at a glance I'm not sure what it does. This doesn't feel like a strict improvement in legibility to me.

Reply:
I don't think that `scan` is well suited for this case because it
- Provides unused functionality
- Implicitly converts `Option<T>` to `T` (`.map(Option::unwrap)` does it explicitly)
- Is rarely used (I think much less people know how this combinator works, if compare with `take_while` or `map`).

However I changed previous approach to `filter_map` with placeholder function, hope it's more understandable.

Waiting for your feedback. 😊

P.S. At least first argument in `scan` closure should be `()` instead of `_` to highlight that internal state isn't used at all.

Reply:
What you're saying makes sense!

Question now though: it seems impls are now half `scan` based, and the other half `take_while` based. Would it make sense to move all of them over to be based on `take_while`?

Reply:
The problem with `Result`s implementation is that it takes first `Err` from provided stream.
However `take_while` accepts reference, while `scan` takes ownership. This can be solved by more complex `filter_map`, which also takes ownership.

Reply:
Or I thought about passing `found_error`'s mut reference as `scan` state and changing it in appropriate way.